File: examples/MovingAverage/client/go/Dynamic/MovingAverageClient.go
package main

import (
    "../gen-go/com/maxeler/MovingAverage"
    "fmt"
    "git.apache.org/thrift.git/lib/go/thrift"
    "os"
    "math/rand"
    "time"
)

func checkForErrors(err error){
    if err != nil {
        fmt.Fprintln(os.Stderr, err)
        os.Exit(-1)
    }
}

func check(dataIn []float64, dataOut []float64, size int32) (int){
    status := 0
    for i := 1; int32(i) < size - 1; i++ {
        a := float32(dataIn[i - 1])
        b := float32(dataIn[i])
        c := float32(dataIn[i + 1])
        res := (a + b + c) / 3
        if dataOut[i] != float64(res) {
            fmt.Printf("Output data @ %v = %v (expected %v)\n", i, dataOut[i], res)
            status++
        }
    }
    
    return status
}

// Calculates time passed from startTime in seconds
func calcTime(startTime time.Time) (float64){
    return float64(time.Now().UnixNano() - startTime.UnixNano()) / 1000000000
}

func main() {
    startTime := time.Now()
    startDFETime := startTime

    // Make socket
    transport, err := thrift.NewTSocket("localhost:9090"); checkForErrors(err)

    // Wrap in a protocol
    protocol := thrift.NewTBinaryProtocolFactoryDefault()
    
    // Create a client to use the protocol encoder
    client := MovingAverage.NewMovingAverageServiceClientFactory(transport, protocol)

    currentTime := calcTime(startTime)
    fmt.Printf("Creating a client:\t\t\t\t%.5fs\n", currentTime)

    // Connect!
    startTime = time.Now()
    err = transport.Open(); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Opening connection:\t\t\t\t%.5fs\n", currentTime)

    startTime = time.Now()
    const size int32 = 384
    sizeBytes := size * 4
    // Generate input data
    dataIn := make([]float64, size)
    for i := 0; int32(i) < size; i++ {
        dataIn[i] = float64(rand.Intn(100))
    }
    currentTime = calcTime(startTime)
    fmt.Printf("Generating input data:\t\t\t\t%.5fs\n", currentTime)

    // Initialize maxfile
    startTime = time.Now()
    maxfile, err := client.MovingAverageInit(); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Initializing maxfile:\t\t\t\t%.5fs\n", currentTime)

    // Load DFE
    startTime = time.Now()
    engine, err := client.MaxLoad(maxfile, "*"); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Loading DFE:\t\t\t\t\t%.5fs\n", currentTime)

    // Allocate and send input streams to server
    startTime = time.Now()
    addressDataIn, err := client.MallocFloat(int64(size)); checkForErrors(err)
    err = client.SendDataFloat(addressDataIn, dataIn); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Sending input data:\t\t\t\t%.5fs\n", currentTime)

    // Allocate memory for output stream on server
    startTime = time.Now()
    addressDataOut, err := client.MallocFloat(int64(size)); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Allocating memory for output stream on server:\t%.5fs\n", currentTime)

    // Action default
    startTime = time.Now()
    actions, err := client.MaxActionsInit(maxfile, "default"); checkForErrors(err)
    err = client.MaxSetParamUint64t(actions, "N", int64(size)); checkForErrors(err)
    err = client.MaxQueueInput(actions, "x", addressDataIn, int64(sizeBytes)); checkForErrors(err)
    err = client.MaxQueueOutput(actions, "y", addressDataOut, int64(sizeBytes)); checkForErrors(err)
    err = client.MaxRun(engine, actions); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Moving average time:\t\t\t\t%.5fs\n", currentTime)

    // Unload DFE
    startTime = time.Now()
    err = client.MaxUnload(engine); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Unloading DFE:\t\t\t\t\t%.5fs\n", currentTime)

    // Get output stream from server
    startTime = time.Now()
    dataOut, err := client.ReceiveDataFloat(addressDataOut, int64(size)); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Getting output stream:\t(size = %v bit)\t%.5fs\n", size * 32, currentTime)

    // Free allocated memory for streams on server
    startTime = time.Now()
    err = client.Free(addressDataIn); checkForErrors(err)
    err = client.Free(addressDataOut); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Freeing allocated memory for streams on server:\t%.5fs\n", currentTime)

    // Free allocated maxfile data
    startTime = time.Now()
    client.MovingAverageFree();
    currentTime = calcTime(startTime)
    fmt.Printf("Freeing allocated maxfile data:\t\t\t%.5fs\n", currentTime)

    // Checking results
    startTime = time.Now()
    status := check(dataIn, dataOut, size)
    currentTime = calcTime(startTime)
    fmt.Printf("Checking results:\t\t\t\t%.5fs\n", currentTime)

    startTime = time.Now()
    defer transport.Close()
    currentTime = calcTime(startTime)
    fmt.Printf("Closing connection:\t\t\t\t%.5fs\n", currentTime)

    currentTime = calcTime(startDFETime)
    fmt.Printf("DFE moving average total time:\t\t\t%.5fs\n", currentTime)

    if (status==0) { 
        fmt.Printf("Test successful!\n")
    } else {
        fmt.Printf("Test failed %v times!", status)
        os.Exit(-1)
    }
}

File: examples/MovingAverage/client/java/AdvancedStatic/MovingAverageClient.java
import com.maxeler.MovingAverage.MovingAverageService;
import com.maxeler.MovingAverage.MovingAverage_actions_t_struct;

import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/** MovingAverage AdvancedStatic example. */
public final class MovingAverageClient {

  /** Average range. */
  private static final int AVERAGE_RANGE = 3;

  /** Minimal random number. */
  private static final int MIN_RAND_NUM = 0;

  /** Maximal random number. */
  private static final int MAX_RAND_NUM = 1000;

  /** Size of int in bits. */
  private static final int SIZE_OF_INT = 32;

  /** Number of nano seconds in one second. */
  private static final int NUM_OF_NANO_SECONDS = 1000000000;

  /** Server port. */
  private static final int PORT = 9090;

  /** Utility classes should not have a public or default constructor. */
  private MovingAverageClient() { }

  /**
   * Generates random data.
   *
   * @param size    Size
   *
   * @return        Random data
   */
  public static List<Double> randomData(final int size) {
    List<Double> randomData = new ArrayList<Double>();

    Random rn = new Random();
    final int range = MAX_RAND_NUM - MIN_RAND_NUM + 1;

    for (int i = 0; i < size; i++) {
      randomData.add((double) rn.nextInt(range) + MIN_RAND_NUM);
    }

    return randomData;
  }

  /**
   * Checks if movingAverageDfe and movingAverageCpu return the same value.
   *
   * @param dataOutDfe  Data output from DFE
   * @param dataOutCpu  Data output from CPU
   * @param size        Size
   */
  public static void check(final List<Double> dataOutDfe,
                           final List<Double> dataOutCpu, final int size) {
    for (int i = 1; i < size - 1; i++) {
      if (!dataOutDfe.get(i).equals(dataOutCpu.get(i))) {
        System.out.println(
            "Output data @ " + i + " = " + dataOutDfe.get(i)
            + " (expected " + dataOutCpu.get(i) + ")");

        System.out.println("Test failed!");
        System.exit(-1);
      }
    }

    System.out.println("Test successful!");
  }

  /**
   * MovingAverage on CPU.
   *
   * @param size    Size
   * @param dataIn  Data input
   *
   * @return        Data output
   */
  public static List<Double> movingAverageCpu(
      final int size, final List<Double> dataIn) {
    List<Double> dataOut = new ArrayList<Double>();

    dataOut.add((dataIn.get(0) + dataIn.get(1)) / 2);

    for (int i = 1; i < size - 1; i++) {
      final float previous = (float) (double) dataIn.get(i - 1);
      final float current = (float) (double) dataIn.get(i);
      final float next = (float) (double) dataIn.get(i + 1);
      final float average = (previous + current + next) / AVERAGE_RANGE;
      dataOut.add((double) average);
    }

    dataOut.add((dataIn.get(size - 1) + dataIn.get(size - 2)) / 2);

    return dataOut;
  }

  /**
   * MovingAverage on DFE.
   *
   * @param size    Size
   * @param dataIn  Data input
   *
   * @return        Data output
   */
  public static List<Double> movingAverageDfe(
      final int size, final List<Double> dataIn) {
    List<Double> dataOut = new ArrayList<Double>();

    DecimalFormat timeFormat = new DecimalFormat("#0.00000");
    double startTime = System.nanoTime();

    // Make socket
    TTransport transport = new TSocket("localhost", PORT);

    // Wrap in a protocol
    TProtocol protocol = new TBinaryProtocol(transport);

    // Create a client to use the protocol encoder
    MovingAverageService.Client client =
        new MovingAverageService.Client(protocol);

    double estimatedTime = (System.nanoTime() - startTime)
                           / NUM_OF_NANO_SECONDS;
    System.out.println("Createing a client:\t\t\t\t\t"
                       + timeFormat.format(estimatedTime) + "s");

    try {
      // Connect!
      startTime = System.nanoTime();
      transport.open();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Opening connection:\t\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Initialize maxfile
      startTime = System.nanoTime();
      final long maxfile = client.MovingAverage_init();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Initializing maxfile:\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Load DFE
      startTime = System.nanoTime();
      final long engine = client.max_load(maxfile, "*");
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Loading DFE:\t\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Allocate and send input streams to server
      startTime = System.nanoTime();
      final long addressDataIn = client.malloc_float(size);
      client.send_data_float(addressDataIn, dataIn);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Sending input data:\t\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Allocate memory for output stream on server
      startTime = System.nanoTime();
      final long addressDataOut = client.malloc_float(size);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Allocating memory for output stream on server:\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Action default
      startTime = System.nanoTime();
      MovingAverage_actions_t_struct action =
          new MovingAverage_actions_t_struct(size, addressDataIn,
                                             addressDataOut);
      final long addressAction = client.send_MovingAverage_actions_t(action);
      client.MovingAverage_run(engine, addressAction);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Moving average time:\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Unload DFE
      startTime = System.nanoTime();
      client.max_unload(engine);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Unloading DFE:\t\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Get output stream from server
      startTime = System.nanoTime();
      dataOut = client.receive_data_float(addressDataOut, size);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Getting output stream:\t(size = "
                         + size * SIZE_OF_INT + " bit)\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Free allocated memory for streams on server
      startTime = System.nanoTime();
      client.free(addressDataIn);
      client.free(addressDataOut);
      client.free(addressAction);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Freeing allocated memory for streams on server:\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Free allocated maxfile data
      startTime = System.nanoTime();
      client.MovingAverage_free();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Freeing allocated maxfile data:\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

      // Close!
      startTime = System.nanoTime();
      transport.close();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Closing connection:\t\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

    } catch (TException x) {
      x.printStackTrace();
      System.exit(-1);
    }

    return dataOut;
  }

  /**
   * Calculates movingAverageDfe and movingAverageCpu and
   * checks if they return the same value.
   *
   * @param args Command line arguments
   */
  public static void main(final String[] args) {
    DecimalFormat timeFormat = new DecimalFormat("#0.00000");
    final int size = 384;

    // Generate data
    double startTime = System.nanoTime();
    final List<Double> dataIn = randomData(size);
    double estimatedTime = (System.nanoTime() - startTime)
                           / NUM_OF_NANO_SECONDS;
    System.out.println("Generating input data:\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

    // DFE Output
    startTime = System.nanoTime();
    final List<Double> dataOutDfe = movingAverageDfe(size, dataIn);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("DFE moving average total time:\t\t\t"
                       + timeFormat.format(estimatedTime) + "s");

    // CPU Output
    startTime = System.nanoTime();
    final List<Double> dataOutCpu = movingAverageCpu(size, dataIn);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("CPU moving average total time:\t\t\t"
                       + timeFormat.format(estimatedTime) + "s");

    // Checking results
    startTime = System.nanoTime();
    check(dataOutDfe, dataOutCpu, size);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("Checking results:\t\t\t\t\t"
                       + timeFormat.format(estimatedTime) + "s");
  }
}

